rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs

```diff
@@ -0,0 +1,112 @@
+use polars::prelude::*;
+use polars_plan::dsl::{Expr, FunctionExpr};
+
+use crate::core::{Function, MetricSpace, StabilityMap, Transformation};
+use crate::domains::{AtomDomain, CategoricalDomain, ExprDomain, OuterMetric};
+use crate::error::*;
+use crate::polars::ExprFunction;
+use crate::transformations::DatasetMetric;
+
+use super::StableExpr;
+
+#[cfg(test)]
+mod test;
+
+/// Make a Transformation that returns a `to_physical` expression for a LazyFrame.
+///
+/// # Arguments
+/// * `input_domain` - Expr domain
+/// * `input_metric` - The metric under which neighboring LazyFrames are compared
+/// * `expr` - The clipping expression
+pub fn make_expr_to_physical<M: OuterMetric>(
+    input_domain: ExprDomain,
+    input_metric: M,
+    expr: Expr,
+) -> Fallible<Transformation<ExprDomain, ExprDomain, M, M>>
+where
+    M::InnerMetric: DatasetMetric,
+    M::Distance: Clone,
+    (ExprDomain, M): MetricSpace,
+    Expr: StableExpr<M, M>,
+{
+    let Expr::Function {
+        input, function, ..
+    } = expr
+    else {
+        return fallible!(MakeTransformation, "expected function expression");
+    };
+
+    if !matches!(function, FunctionExpr::ToPhysical) {
+        return fallible!(
+            MakeTransformation,
+            "expected to_physical function, found {}",
+            function
+        );
+    };
+
+    let n_args = input.len();
+    let [input] = <[Expr; 1]>::try_from(input).map_err(|_| {
+        err!(
+            MakeTransformation,
+            "to_physical expects 1 data argument, found {}",
+            n_args
+        )
+    })?;
+
+    let t_prior = input.make_stable(input_domain.clone(), input_metric.clone())?;
+    let (middle_domain, middle_metric) = t_prior.output_space();
+
+    let mut output_domain = middle_domain.clone();
+
+    let active_series = output_domain.active_series_mut()?;
+
+    use DataType::*;
+    let in_dtype = &active_series.field.dtype;
+
+    // this code is written intentionally to fail or change if polars behavior changes
+    match (in_dtype.clone(), in_dtype.to_physical()) {
+        (in_dtype, out_dtype) if in_dtype == out_dtype => (),
+        (Categorical(_, _), UInt32) => {
+            let cat_domain = active_series
+                .element_domain
+                .as_any()
+                .downcast_ref::<CategoricalDomain>()
+                .ok_or_else(|| {
+                    err!(
+                        MakeTransformation,
+                        "to_physical: expected categorical series domain"
+                    )
+                })?;
+
+            if !cat_domain.encoding().is_some() {
+                return fallible!(MakeTransformation, "to_physical: to prevent potentially revealing information about row ordering, category ordering must be statically known");
```

Comment:
If the python interface has an encoding kwarg, should we mention `encoding` in this message to steer the user towards a fix?

Reply:
This is something an enum dtype would make possible, but I can't steer users there yet.